File: packages/server/src/args.ts
/** Argument validation and coercion at the system boundary. */
import { typeRefToString, type Annotation, type ArgDef, type FieldDef, type RayfoldSchemaIR, type TypeRef } from "@rayfold/schema";
import { RayfoldError } from "./protocol.ts";

const MAX_PAGE_FIRST = 200;
const NUMERIC = new Set(["Int", "Long", "Float", "Decimal"]);

export function coerceArgs(ir: RayfoldSchemaIR, defs: ArgDef[], raw: unknown, path: string): Record<string, unknown> {
  if (raw !== undefined && (raw === null || typeof raw !== "object" || Array.isArray(raw))) {
    throw new RayfoldError("invalid_argument", `${path}: expected an object`);
  }
  const input = (raw ?? {}) as Record<string, unknown>;
  for (const k of Object.keys(input)) {
    if (!defs.some((d) => d.name === k)) throw new RayfoldError("invalid_argument", `${path}.${k}: unknown argument`);
  }
  const out: Record<string, unknown> = {};
  for (const d of defs) {
    const v = input[d.name];
    const p = `${path}.${d.name}`;
    if (v === undefined) {
      if (d.default !== undefined) {
        out[d.name] = coerceValue(ir, d.type, d.default, p);
        continue;
      }
      if (!d.type.nullable) throw new RayfoldError("invalid_argument", `${p}: required`);
      continue; // absent stays absent: partial updates (PATCH) depend on the difference from null
    }
    if (v === null) {
      // An explicit null is never replaced by the default: the caller asked for null.
      if (!d.type.nullable) throw new RayfoldError("invalid_argument", d.default === undefined ? `${p}: required` : `${p}: must not be null`);
      out[d.name] = null;
      continue;
    }
    out[d.name] = coerceValue(ir, d.type, v, p);
    checkConstraints(d.annotations, d.type, out[d.name], p);
  }
  return out;
}

/**
 * Declarative value constraints from the schema (spec 01 section 4): `@range(min, max)` on numbers and
 * on string/list lengths. Enforced before any resolver runs, so a bad value never causes a side effect.
 */
export function checkConstraints(annotations: Annotation[], t: TypeRef, v: unknown, path: string): void {
  if (v === null || v === undefined) return;
  const range = annotations.find((a) => a.name === "range");
  if (range) {
    const min = typeof range.args["min"] === "number" ? range.args["min"] : undefined;
    const max = typeof range.args["max"] === "number" ? range.args["max"] : undefined;
    // The declared type decides: a String "1984" is four characters long, a Decimal "10.50" is a value.
    const n =
      t.kind === "list" ? (Array.isArray(v) ? v.length : undefined)
      : NUMERIC.has(t.name) ? (typeof v === "number" ? v : typeof v === "string" ? Number(v) : undefined)
      : typeof v === "string" ? v.length
      : undefined;
    if (n !== undefined) {
      if (min !== undefined && n < min) throw new RayfoldError("invalid_argument", `${path}: must be >= ${min}`);
      if (max !== undefined && n > max) throw new RayfoldError("invalid_argument", `${path}: must be <= ${max}`);
    }
  }
  const fmt = annotations.find((a) => a.name === "format");
  const pattern = fmt?.args["pattern"];
  if (typeof pattern === "string" && typeof v === "string" && !new RegExp(pattern).test(v)) throw new RayfoldError("invalid_argument", `${path}: must match ${pattern}`);
}

export function coerceValue(ir: RayfoldSchemaIR, t: TypeRef, v: unknown, path: string): unknown {
  if (v === null || v === undefined) {
    if (t.nullable) return null;
    throw new RayfoldError("invalid_argument", `${path}: must not be null`);
  }
  if (t.kind === "list") {
    if (!Array.isArray(v)) throw new RayfoldError("invalid_argument", `${path}: expected a list`);
    return v.map((x, i) => coerceValue(ir, t.of, x, `${path}.${i}`));
  }
  const def = ir.types[t.name];
  if (!def) throw new RayfoldError("internal", `${path}: unknown type ${t.name}`);
  switch (def.kind) {
    case "scalar":
      return coerceScalar(t.name, v, path);
    case "enum":
      if (typeof v !== "string" || !def.values.some((x) => x.name === v)) {
        throw new RayfoldError("invalid_argument", `${path}: expected one of ${def.values.map((x) => x.name).join(", ")}`);
      }
      return v;
    case "input": {
      if (typeof v !== "object" || Array.isArray(v)) throw new RayfoldError("invalid_argument", `${path}: expected ${t.name}`);
      const obj = coerceArgs(ir, def.fields.map(fieldAsArg), v, path);
      if (t.name === "PageArgs") {
        const first = obj["first"];
        if (typeof first === "number" && first > MAX_PAGE_FIRST) obj["first"] = MAX_PAGE_FIRST;
        if (typeof first === "number" && first < 0) throw new RayfoldError("invalid_argument", `${path}.first: must be >= 0`);
        const offset = obj["offset"];
        if (typeof offset === "number" && offset < 0) throw new RayfoldError("invalid_argument", `${path}.offset: must be >= 0`);
      }
      return obj;
    }
    default:
      throw new RayfoldError("invalid_argument", `${path}: ${typeRefToString(t)} is not an input type`);
  }
}

function fieldAsArg(f: FieldDef): ArgDef {
  const a: ArgDef = { name: f.name, type: f.type, annotations: f.annotations };
  if (f.default !== undefined) a.default = f.default;
  return a;
}

const RFC3339 = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$/;
const DATE = /^\d{4}-\d{2}-\d{2}$/;
const DECIMAL = /^-?\d+(\.\d+)?$/;

export function coerceScalar(name: string, v: unknown, path: string): unknown {
  const bad = (want: string): never => {
    throw new RayfoldError("invalid_argument", `${path}: expected ${want}`);
  };
  switch (name) {
    case "ID":
      if (typeof v === "string" && v.length > 0) return v;
      if (typeof v === "number" && Number.isSafeInteger(v)) return String(v);
      return bad("ID");
    case "String":
      return typeof v === "string" ? v : bad("String");
    case "Int":
      return typeof v === "number" && Number.isInteger(v) && Math.abs(v) <= 2_147_483_647 ? v : bad("Int");
    case "Long":
      if (typeof v === "number" && Number.isSafeInteger(v)) return v; // larger values lose digits as JSON numbers: send them as text
      if (typeof v === "string" && /^-?\d+$/.test(v)) return v;
      return bad("Long");
    case "Float":
      return typeof v === "number" && Number.isFinite(v) ? v : bad("Float");
    case "Boolean":
      return typeof v === "boolean" ? v : bad("Boolean");
    case "Decimal":
      if (typeof v === "string" && DECIMAL.test(v)) return v;
      if (typeof v === "number" && Number.isFinite(v) && DECIMAL.test(String(v))) return String(v); // 1e21 would become "1e+21"
      return bad("Decimal");
    case "Instant":
      return typeof v === "string" && RFC3339.test(v) ? v : bad("Instant (RFC 3339)");
    case "Date":
      return typeof v === "string" && DATE.test(v) ? v : bad("Date (YYYY-MM-DD)");
    case "Duration":
      if (typeof v === "number" && v >= 0) return v;
      if (typeof v === "string" && /^\d+(ms|s|m|h|d)$/.test(v)) return v;
      return bad("Duration");
    case "Bytes":
      return typeof v === "string" && /^[A-Za-z0-9_-]*$/.test(v) ? v : bad("Bytes (base64url)");
    case "JSON":
      return v;
    default:
      // user-defined scalar: accept any JSON scalar
      return typeof v === "object" ? bad(`scalar ${name}`) : v;
  }
}

/** Replace `{ "$ref": "id.path" }` objects with values from earlier results. Spec 03 §2. */
export function resolveRefs(value: unknown, lookup: (opId: number, path: string[]) => unknown, at: string): unknown {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map((x, i) => resolveRefs(x, lookup, `${at}.${i}`));
  const obj = value as Record<string, unknown>;
  const ref = obj["$ref"];
  if (typeof ref === "string" && Object.keys(obj).length === 1) {
    const [idText, ...path] = ref.split(".");
    const id = Number(idText);
    if (!Number.isInteger(id) || id <= 0) throw new RayfoldError("invalid_argument", `${at}: bad $ref ${JSON.stringify(ref)}`);
    const v = lookup(id, path);
    if (v === undefined) throw new RayfoldError("invalid_argument", `${at}: $ref ${ref} resolved to nothing`);
    return v;
  }
  const out: Record<string, unknown> = {};
  // an own property, as JSON.parse made it: assigning "__proto__" would replace the copy's prototype and smuggle in arguments
  for (const [k, v] of Object.entries(obj)) Object.defineProperty(out, k, { value: resolveRefs(v, lookup, `${at}.${k}`), enumerable: true, writable: true, configurable: true });
  return out;
}

/** Ids of ops referenced by `$ref` anywhere inside `value`. */
export function collectRefs(value: unknown, out = new Set<number>()): Set<number> {
  if (value === null || typeof value !== "object") return out;
  if (Array.isArray(value)) {
    value.forEach((x) => collectRefs(x, out));
    return out;
  }
  const obj = value as Record<string, unknown>;
  const ref = obj["$ref"];
  if (typeof ref === "string" && Object.keys(obj).length === 1) {
    out.add(Number(ref.split(".")[0]));
    return out;
  }
  for (const v of Object.values(obj)) collectRefs(v, out);
  return out;
}

export function getPath(value: unknown, path: string[]): unknown {
  let cur = value;
  for (const p of path) {
    if (cur === null || cur === undefined || typeof cur !== "object") return undefined;
    if (Array.isArray(cur)) {
      cur = cur[Number(p)];
      continue;
    }
    // own data only: "__proto__", "constructor" and friends are never reachable through a $ref path
    if (!Object.prototype.hasOwnProperty.call(cur, p)) return undefined;
    cur = (cur as Record<string, unknown>)[p];
  }
  return cur;
}

File: kotlin/rayfold-core/src/main/kotlin/dev/rayfold/core/Args.kt
package dev.rayfold.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

/** Argument validation and coercion at the system boundary (mirrors packages/server/src/args.ts). */
object Args {
    internal const val MAX_PAGE_FIRST = 200
    private const val MAX_FORMAT_INPUT = 10_000

    /**
     * [wire] reads each member under its wire name (`@http(name:)`, spec 04 section 8), as an HTTP binding receives it, and
     * names it that way in errors; the result is keyed by schema names either way.
     */
    fun coerce(ir: RayfoldSchemaIR, defs: List<ArgDef>, raw: JsonElement?, path: String, returns: TypeRef? = null, wire: Boolean = false): JsonObject {
        fun key(d: ArgDef) = if (wire) d.wireName else d.name
        if (raw != null && raw !is JsonNull && raw !is JsonObject) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: expected an object")
        val input = raw as? JsonObject ?: JsonObject(emptyMap())
        for (k in input.keys) if (defs.none { key(it) == k }) throw RayfoldException(Code.INVALID_ARGUMENT, "$path.$k: unknown argument")
        val out = linkedMapOf<String, JsonElement>()
        for (d in defs) {
            val v = input[key(d)]
            val p = "$path.${key(d)}"
            if (v == null) {
                if (d.default != null) { out[d.name] = value(ir, d.type, d.default, p); continue }
                if (!d.type.nullable) throw RayfoldException(Code.INVALID_ARGUMENT, "$p: required")
                continue // absent stays absent, explicit null stays null: partial updates depend on the difference
            }
            if (v is JsonNull) {
                // a default fills only an absent value; it never masks an explicit null
                if (d.type.nullable) { out[d.name] = JsonNull; continue }
                throw RayfoldException(Code.INVALID_ARGUMENT, if (d.default != null) "$p: must not be null" else "$p: required")
            }
            val coerced = value(ir, d.type, v, p, wire)
            out[d.name] = coerced
            checkConstraints(ir, d.annotations, coerced, p, d.type)
        }
        // spec 01 section 6: a page may take `first` as an argument of its own rather than inside PageArgs, and is capped the same
        val first = (out["first"] as? JsonPrimitive)?.takeIf { returns?.isPage == true && !it.isString }?.contentOrNull?.toDoubleOrNull()
        if (first != null && first < 0) throw RayfoldException(Code.INVALID_ARGUMENT, "$path.first: must be >= 0")
        if (first != null && first > MAX_PAGE_FIRST) out["first"] = JsonPrimitive(MAX_PAGE_FIRST)
        return JsonObject(out)
    }

    private val NUMERIC = setOf("Int", "Long", "Float", "Decimal")

    /**
     * `@range(min, max)` on numbers and string/list lengths, `@format(pattern:)` on strings (spec 01 section 4): the
     * whole string must match, so `$` cannot be satisfied by a trailing line break.
     * Range goes by the declared type, not by how the value looks: a String "1984" is 4 characters, a Decimal "1984" is 1984.
     */
    fun checkConstraints(ir: RayfoldSchemaIR, annotations: List<Annotation>, v: JsonElement, path: String, type: TypeRef) {
        if (v is JsonNull) return
        annotations.find("range")?.let { r ->
            val min = r.args["min"].numberOrNull()
            val max = r.args["max"].numberOrNull()
            val n: Double? = when (v) {
                is JsonArray -> v.size.toDouble()
                is JsonPrimitive -> when {
                    !type.isList && type.name in NUMERIC -> v.contentOrNull?.toDoubleOrNull()
                    v.isString -> v.content.length.toDouble()
                    else -> v.contentOrNull?.toDoubleOrNull()
                }
                else -> null
            }
            if (n != null) {
                if (min != null && n < min) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: must be >= ${fmtNum(min)}")
                if (max != null && n > max) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: must be <= ${fmtNum(max)}")
            }
        }
        val pattern = (annotations.find("format")?.args?.get("pattern") as? JsonPrimitive)?.takeIf { it.isString }?.content
        if (pattern != null && v is JsonPrimitive && v.isString) {
            val s = v.content
            // bounded before matching: backtracking cost and recursion depth grow with the input
            if (s.length > MAX_FORMAT_INPUT) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: longer than $MAX_FORMAT_INPUT characters, too long to match $pattern")
            val ok = try {
                ir.formatRegex(pattern).matches(s)
            } catch (e: StackOverflowError) {
                throw RayfoldException(Code.INVALID_ARGUMENT, "$path: too complex to match $pattern")
            }
            if (!ok) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: must match $pattern")
        }
    }

    private fun fmtNum(d: Double): String = if (d == Math.floor(d) && !d.isInfinite()) d.toLong().toString() else d.toString()

    fun value(ir: RayfoldSchemaIR, t: TypeRef, v: JsonElement, path: String, wire: Boolean = false): JsonElement {
        if (v is JsonNull) {
            if (t.nullable) return JsonNull
            throw RayfoldException(Code.INVALID_ARGUMENT, "$path: must not be null")
        }
        if (t.isList) {
            val arr = v as? JsonArray ?: throw RayfoldException(Code.INVALID_ARGUMENT, "$path: expected a list")
            return JsonArray(arr.mapIndexed { i, x -> value(ir, t.element, x, "$path.$i", wire) })
        }
        val def = ir.types[t.name] ?: throw RayfoldException(Code.INTERNAL, "$path: unknown type ${t.name}")
        return when (def.kind) {
            "scalar" -> scalar(t.typeName, v, path)
            "enum" -> {
                val s = (v as? JsonPrimitive)?.takeIf { it.isString }?.content
                if (s == null || def.values.none { it.name == s }) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: expected one of ${def.values.joinToString(", ") { it.name }}")
                v
            }
            "input" -> {
                if (v !is JsonObject) throw RayfoldException(Code.INVALID_ARGUMENT, "$path: expected ${t.name}")
                val obj = coerce(ir, def.fields.map { ArgDef(it.name, it.description, it.type, it.default, it.annotations) }, v, path, wire = wire).toMutableMap()
                if (t.name == "PageArgs") {
                    val first = (obj["first"] as? JsonPrimitive)?.contentOrNull?.toDoubleOrNull()
                    if (first != null && first > MAX_PAGE_FIRST) obj["first"] = JsonPrimitive(MAX_PAGE_FIRST)
                    if (first != null && first < 0) throw RayfoldException(Code.INVALID_ARGUMENT, "$path.first: must be >= 0")
                    val offset = (obj["offset"] as? JsonPrimitive)?.contentOrNull?.toDoubleOrNull()
                    if (offset != null && offset < 0) throw RayfoldException(Code.INVALID_ARGUMENT, "$path.offset: must be >= 0")
                }
                JsonObject(obj)
            }
            else -> throw RayfoldException(Code.INVALID_ARGUMENT, "$path: $t is not an input type")
        }
    }

    private val RFC3339 = Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?(Z|[+-]\\d{2}:\\d{2})$")
    private val DATE = Regex("^\\d{4}-\\d{2}-\\d{2}$")
    private val DECIMAL = Regex("^-?\\d+(\\.\\d+)?$")
    private val LONG_TEXT = Regex("^-?\\d+$")

    fun scalar(name: String, v: JsonElement, path: String): JsonElement {
        val p = v as? JsonPrimitive ?: return if (name == "JSON") v else bad(path, "scalar $name")
        val str = p.takeIf { it.isString }?.content
        // strict JSON numbers only: NaN, Infinity, 1d and 01 are not numbers, whatever a lenient parser let through
        val num = if (!p.isString && StrictJson.isNumber(p.content)) p.content.toDoubleOrNull()?.takeIf { it.isFinite() } else null
        val isInt = num != null && StrictJson.integerOrNull(p) != null
        return when (name) {
            "ID" -> if (str != null && str.isNotEmpty()) p else if (isInt) JsonPrimitive(p.content) else bad(path, "ID")
            "String" -> if (str != null) p else bad(path, "String")
            "Int" -> if (isInt && p.content.toIntOrNull() != null) p else bad(path, "Int")
            "Long" -> if ((isInt && p.content.toLongOrNull() != null) || (str != null && LONG_TEXT.matches(str) && str.toLongOrNull() != null)) p else bad(path, "Long")
            "Float" -> if (num != null) p else bad(path, "Float")
            "Boolean" -> if (!p.isString && (p.content == "true" || p.content == "false")) p else bad(path, "Boolean")
            "Decimal" -> if (str != null && DECIMAL.matches(str)) p else if (num != null) JsonPrimitive(p.content) else bad(path, "Decimal")
            "Instant" -> if (str != null && RFC3339.matches(str)) p else bad(path, "Instant (RFC 3339)")
            "Date" -> if (str != null && DATE.matches(str)) p else bad(path, "Date (YYYY-MM-DD)")
            "Duration" -> if ((num != null && num >= 0) || (str != null && Regex("^\\d+(ms|s|m|h|d)$").matches(str))) p else bad(path, "Duration")
            "Bytes" -> if (str != null && Regex("^[A-Za-z0-9_-]*$").matches(str)) p else bad(path, "Bytes (base64url)")
            "JSON" -> v
            else -> p
        }
    }

    private fun bad(path: String, want: String): Nothing = throw RayfoldException(Code.INVALID_ARGUMENT, "$path: expected $want")

    /** Replace `{ "$ref": "id.path" }` with values from earlier results (spec/03 section 2). */
    fun resolveRefs(value: JsonElement, lookup: (Int, List<String>) -> JsonElement?, at: String): JsonElement = when (value) {
        is JsonArray -> JsonArray(value.mapIndexed { i, x -> resolveRefs(x, lookup, "$at.$i") })
        is JsonObject -> {
            val ref = (value["\$ref"] as? JsonPrimitive)?.takeIf { it.isString }?.content
            if (ref != null && value.size == 1) {
                val parts = ref.split(".")
                val id = parts[0].toIntOrNull()
                if (id == null || id <= 0) throw RayfoldException(Code.INVALID_ARGUMENT, "$at: bad \$ref \"$ref\"")
                lookup(id, parts.drop(1)) ?: throw RayfoldException(Code.INVALID_ARGUMENT, "$at: \$ref $ref resolved to nothing")
            } else JsonObject(value.mapValues { (k, v) -> resolveRefs(v, lookup, "$at.$k") })
        }
        else -> value
    }

    /** `{ "$ref": "..." }`: a value known only once an earlier op has run. */
    fun isRef(v: JsonElement): Boolean = v is JsonObject && v.size == 1 && (v["\$ref"] as? JsonPrimitive)?.isString == true

    fun collectRefs(value: JsonElement, out: MutableSet<Int> = linkedSetOf()): Set<Int> {
        when (value) {
            is JsonArray -> value.forEach { collectRefs(it, out) }
            is JsonObject -> {
                val ref = (value["\$ref"] as? JsonPrimitive)?.takeIf { it.isString }?.content
                if (ref != null && value.size == 1) out.add(ref.substringBefore(".").toIntOrNull() ?: -1)
                else value.values.forEach { collectRefs(it, out) }
            }
            else -> {}
        }
        return out
    }

    fun getPath(value: JsonElement?, path: List<String>): JsonElement? {
        var cur = value
        for (p in path) {
            cur = when (cur) {
                is JsonObject -> cur[p]
                is JsonArray -> cur.getOrNull(p.toIntOrNull() ?: return null)
                else -> return null
            } ?: return null
        }
        return cur
    }
}
